### PR TITLE
feat(price-feeder): computed price api routes

### PR DESCRIPTION
**Pull request by zarazan**
## Description

I did some performance measurements before and after this change and the timing difference is negligible. That said, I think we could get some performance improvements by computing these per-provider prices first and passing them to the convert/filter/compute functions however it will require a decent amount of re-factoring that would make this PR much larger.

- Added additional API routes for the price-feeder to debug the per-provider computed prices
- Remove unused error handling from ComputeVWAP
- Updated the lock on the Oracle.prices variable to lock when writing

closes: https://github.com/umee-network/umee/issues/1263#event-7378865002

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] added appropriate labels to the PR
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: price-feeder/oracle/oracle.go [hunk 1]
```diff
@@ -163,6 +168,20 @@ func (o *Oracle) GetPrices() map[string]sdk.Dec {
...
+	defer o.tvwapsMutex.RUnlock()
+	return o.tvwapsByProvider
+}
+
+// GetVwapPrices returns the vwapsByProvider map using a read lock
+func (o *Oracle) GetVwapPrices() map[provider.Name]map[string]sdk.Dec {
+	o.vwapsMutex.RLock()
+	defer o.vwapsMutex.RUnlock()
+	return o.vwapsByProvider
+}
```
Comment (RafilxTenfen) [inline price-feeder/oracle/oracle.go:183]: What do you think about creating a structure to handle mutex for `map[provider.Name]map[string]sdk.Dec` struct?

File: price-feeder/oracle/util.go [hunk 2]
```diff
@@ -184,3 +184,31 @@ func StandardDeviation(
 
 	return deviations, means, nil
 }
+
+// ComputeTvwapsByProvider computes the tvwap prices from candles for each provider separately and returns them
+// in a map separated by provider name
+func ComputeTvwapsByProvider(prices provider.AggregatedProviderCandles) (map[provider.Name]map[string]sdk.Dec, error) {
```
Comment (RafilxTenfen) [inline price-feeder/oracle/util.go:190]: :100:

File: price-feeder/oracle/oracle.go [hunk 3]
```diff
@@ -163,6 +168,20 @@ func (o *Oracle) GetPrices() map[string]sdk.Dec {
 	return prices
 }
 
+// GetTvwapPrices returns the tvwapsByProvider map using a read lock
+func (o *Oracle) GetTvwapPrices() map[provider.Name]map[string]sdk.Dec {
+	o.tvwapsMutex.RLock()
+	defer o.tvwapsMutex.RUnlock()
+	return o.tvwapsByProvider
+}
```
[the remark this answers is not in the document]
Comment (zarazan) [reply price-feeder/oracle/oracle.go:176]: That is true, this is not entirely thread safe. Ideally I wanted to release the read lock after serializing the data for the http response but that would get a little messy. I  think you're right that creating a copy of the nested map is the best option and not that much overhead. I'll update.

File: price-feeder/oracle/oracle.go (hunk 1 above)
Comment (zarazan) [reply price-feeder/oracle/oracle.go:183]: hm, good question. I'm not super against it but my thinking was that they are already in the oracle struct and if I did that we would probably want to do the same thing with pricesMutex and prices. I'll make some changes and see how it looks.

Comment (zarazan) [reply price-feeder/oracle/oracle.go]: https://github.com/umee-network/umee/pull/1445/commits/61664c8b6373a2f01cc65bb4584e8ca4e820eb26

Comment (RafilxTenfen) [reply price-feeder/oracle/oracle.go]: I was trying to suggest something like this #1465

File: price-feeder/oracle/oracle_test.go [hunk 4]
```diff
@@ -467,19 +467,18 @@ func TestSuccessGetComputedPricesCandles(t *testing.T) {
...
+	prices, err := ots.oracle.GetComputedPrices(
 		providerCandles,
 		make(provider.AggregatedProviderPrices, 1),
 		providerPair,
 		make(map[string]sdk.Dec),
 	)
 
-	require.NoError(t, err, "It should successfully get computed candle prices")
-	require.Equal(t, prices[pair.Base], atomPrice)
+	require.NoError(ots.T(), err, "It should successfully get computed candle prices")
```
[the remark this answers is not in the document]
Comment (zarazan) [reply price-feeder/oracle/oracle_test.go:477]: I feel like it is more readable the other way and that seems to be what the docs recommend https://github.com/stretchr/testify#suite-package

File: price-feeder/oracle/oracle.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zarazan) [reply price-feeder/oracle/oracle.go]: https://github.com/umee-network/umee/pull/1445/commits/9f0508d19ce27facc34c79bb0ec82ec95438518f

File: price-feeder/oracle/prices.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/umee-network/umee/price-feeder/oracle/provider"
+)
+
+type (
+	PricesByProvider map[provider.Name]map[string]sdk.Dec
+
+	PricesWithMutex struct {
```
Comment (RafilxTenfen) [inline price-feeder/oracle/prices.go:14]: :100:

**Pull request merged**